File: proto-auth-app.js
const express = require('express');
const bodyParser = require('body-parser');

const app = express();

app.use(bodyParser.json());

const validUsername = "user123";
const validPassword = "pass123";

// Home route (could be a login page)
app.get('/', (req, res) => {
    res.send("Please log in by sending a POST request to /login with your username and password.");
});

// Authentication route
app.post('/login', (req, res) => {
    const { username, password } = req.body;

    if (username === validUsername && password === validPassword) {
        res.status(200).json({ message: "Access Granted. Welcome!" });
    } else {
        res.status(401).json({ message: "Access Denied. Invalid username or password." });
    }
});

app.get('/protected', (req, res) => {
    const authHeader = req.headers.authorization;

    if (!authHeader) {
        return res.status(401).json({ message: "No credentials sent!" });
    }

    // Decode the authorization header (Basic auth)
    const base64Credentials = authHeader.split(' ')[1];
    const credentials = Buffer.from(base64Credentials, 'base64').toString('ascii');
    const [username, password] = credentials.split(':');

    if (username === validUsername && password === validPassword) {
        res.json({ message: "This is a protected endpoint!" });
    } else {
        res.status(401).json({ message: "Access Denied. Invalid credentials." });
    }
});

const PORT = process.env.PORT || 3000;
app.listen(PORT, () => {
    console.log(`Server is running on port ${PORT}`);
});

File: proto-auth-app.py
from flask import Flask, request, jsonify

app = Flask(__name__)

valid_username = "user123"
valid_password = "pass123"

@app.route('/')
def home():
    return "Please log in by sending a POST request with your username and password."

@app.route('/login', methods=['POST'])
def login():
    data = request.get_json()
    username = data.get('username')
    password = data.get('password')

    if username == valid_username and password == valid_password:
        return jsonify({"message": "Access Granted. Welcome!"}), 200
    else:
        return jsonify({"message": "Access Denied. Invalid username or password."}), 401

@app.route('/protected')
def protected():
    auth = request.authorization
    if auth and auth.username == valid_username and auth.password == valid_password:
        return jsonify({"message": "This is a protected endpoint!"})
    return jsonify({"message": "Access Denied. Invalid credentials."}), 401

if __name__ == '__main__':
    app.run(debug=True)
